src/engine/GroupBy.cpp

```diff
@@ -1021,8 +1022,32 @@ void GroupBy::createResultFromHashMap(
     for (auto& alias : aggregateAliases) {
       auto& info = alias.aggregateInfo_;
 
-      // Only one aggregate, and it is at the top of the alias.
-      if (info.size() == 1 && !info.at(0).parentAndIndex_.has_value()) {
+      // Check if the grouped variable occurs in this expression
+      auto groupByVariableSubstitutions =
+          findGroupedVariable(alias.expr_.getPimpl());
+
+      if (groupByVariableSubstitutions.topLevel_) {
+        // If the aggregate is at the top of the alias, e.g. `SELECT (?a as ?x)
+        // WHERE {...} GROUP BY ?a`, we can copy values directly from the column
```

Comment:
This whole substitution code has become so long that it can be a separate function.